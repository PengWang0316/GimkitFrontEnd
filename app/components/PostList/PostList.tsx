import React, {
  memo, useEffect, useCallback, useState,
} from 'react';
import { connect } from 'react-redux';
import { Typography, Paper } from '@material-ui/core';
import I18n from '@kevinwang0316/i18n';
import { makeStyles } from '@material-ui/core/styles';
import Pagination from '@kevinwang0316/react-materialui-pagination';

import ReadDialog from './Dialogs/ReadDialog';
import PostListItem from './PostListItem';
import { AppState } from '../../store/ConfigureStore';
import { PostsType, PostType } from '../../store/Posts/types';
import { PostCountType } from '../../store/PostCount/types';
import { fetchPosts as fetchPostsAction } from '../../store/Posts/actions';
import { fetchPostCount as fetchPostCountAction } from '../../store/PostCount/actions';
import { MAX_POSTS_AMOUNT } from '../../config';

interface Props {
  posts: PostsType;
  postCount: PostCountType;
  fetchPosts: (offset: number) => void;
  fetchPostCount: Function;
}

const useStyles = makeStyles({
  rootPaper: {
    width: '100%',
    padding: 15,
  },
});

export const PostList = ({
  posts = null, postCount = null, fetchPosts, fetchPostCount,
}: Props) => {
  const [isOpenRead, setIsOpenRead] = useState(false);
  const [currentPost, setCurrentPost] = useState(null);
  const [offset, setOffset] = useState(0);
  const classes = useStyles({});

  // Initial fetching
  useEffect(() => {
    if (!posts) fetchPosts(0);
    if (postCount === null) fetchPostCount();
  });

  // callback for the reading dialog display
  const handleReadingClose = () => setIsOpenRead(state => !state);

  // callback for click each item
  const itemClickCallback = useCallback((post: PostType) => {
    setCurrentPost(post);
    handleReadingClose();
  }, []);

  // callback for the pagination button click
  const handlePaginationClick = useCallback((newOffset: number) => {
    console.log(newOffset);
    setOffset(newOffset);
  }, []);

  return (
    <div className={classes.rootPaper}>
      <Typography variant="h5" component="h4">
        {I18n.get('postListTitle')}
      </Typography>
      {posts && posts.map(post => (
        <PostListItem
          key={post.id}
          post={post}
          handleClick={itemClickCallback}
        />
      ))}
      {postCount && (
        <Pagination
          offset={offset}
          limit={MAX_POSTS_AMOUNT}
          total={postCount}
          onClick={handlePaginationClick}
        />
      )}
      <ReadDialog
        post={currentPost}
        isOpen={isOpenRead}
        handleClose={handleReadingClose}
      />
    </div>
  );
};

/* istanbul ignore next */
const mapStateToProps = ({ posts, postCount }: AppState) => ({ posts, postCount });
/* istanbul ignore next */
const mapDispatchToProps = { fetchPostCount: fetchPostCountAction, fetchPosts: fetchPostsAction };
export default connect(mapStateToProps, mapDispatchToProps)(memo(PostList));
