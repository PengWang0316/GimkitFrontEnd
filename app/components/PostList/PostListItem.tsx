import React from 'react';
import {
  List, ListItem, Divider, ListItemText, Typography, IconButton,
} from '@material-ui/core';
import { Delete as DeleteIcon, Edit as EditIcon } from '@material-ui/icons';
import { makeStyles } from '@material-ui/core/styles';
import { pink, blueGrey } from '@material-ui/core/colors';
import parse from 'html-react-parser';

import { PostType } from '../../store/Posts/types';

interface Props {
  post: PostType;
  handleClick: (post: PostType) => void;
  handleDeleteClick: (postId: number) => void;
  handleEditCallback: (post: PostType) => void;
}

const useStyles = makeStyles({
  rootPaper: {
    width: '100%',
    padding: 15,
  },
  list: {
    width: '100%',
  },
  inline: {
    display: 'inline',
  },
  inlineBold: {
    display: 'inline',
    fontWeight: 'bold',
  },
  itemText: {
    maxHeight: 200,
    overflow: 'hidden',
  },
  deleteBtn: {
    marginLeft: 5,
    color: pink[400],
  },
  editBtn: {
    marginLeft: 15,
    color: blueGrey[400],
  },
  icon: {
    width: 20,
    height: 20,
  },
});
const MAX_CONTENT_LENGTH = 300;
const turncate = (text: string): string => text.length < MAX_CONTENT_LENGTH ? text : `${text.slice(0, MAX_CONTENT_LENGTH)}...`;

export const PostListItem = ({
  post, handleClick, handleDeleteClick, handleEditCallback,
}: Props) => {
  const classes = useStyles({});
  const handleDelete = (event: React.MouseEvent) => {
    event.stopPropagation();
    handleDeleteClick(post.id);
  };

  const handleEdit = (event: React.MouseEvent) => {
    event.stopPropagation();
    handleEditCallback(post);
  };

  return (
    <List className={classes.list}>
      <ListItem alignItems="center" button onClick={() => handleClick(post)}>
        <ListItemText
          className={classes.itemText}
          primary={(
            <>
              <Typography
                component="span"
                variant="h6"
                className={classes.inlineBold}
                color="textPrimary"
              >
                {`${post.title} - ${new Date(post.date).toLocaleDateString()}`}
              </Typography>
              <IconButton size="small" className={classes.editBtn} onClick={handleEdit}>
                <EditIcon className={classes.icon} />
              </IconButton>
              <IconButton size="small" className={classes.deleteBtn} onClick={handleDelete}>
                <DeleteIcon className={classes.icon} />
              </IconButton>
            </>
          )}
          secondary={(
            <>
              <Typography
                component="span"
                variant="body2"
                className={classes.inline}
                color="textPrimary"
              >
                <div>{parse(turncate(post.content))}</div>
              </Typography>
            </>
          )}
        />
      </ListItem>
      <Divider component="li" />
    </List>
  );
};

export default PostListItem;
