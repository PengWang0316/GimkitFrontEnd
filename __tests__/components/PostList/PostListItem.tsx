import React from 'react';
import renderer from 'react-test-renderer';

import { PostListItem } from '../../../app/components/PostList/PostListItem';

jest.mock('@material-ui/core/List', () => 'List');
jest.mock('@material-ui/core/ListItem', () => 'ListItem');
jest.mock('@material-ui/core/Divider', () => 'Divider');
jest.mock('@material-ui/core/ListItemText', () => 'ListItemText');
jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/core/IconButton', () => 'IconButton');
jest.mock('@material-ui/icons/Delete', () => 'DeleteIcon');
jest.mock('@material-ui/icons/Edit', () => 'EditIcon');
jest.mock('@kevinwang0316/i18n', () => ({ get: jest.fn().mockReturnValue('I18n text') }));
jest.mock('html-react-parser', () => jest.fn().mockReturnValue('Parsed value'));

describe('PostListItem', () => {
  const defaultProps = {
    post: {
      id: 1, content: 'content', title: 'title', date: '01/01/2019',
    },
    handleClick: jest.fn(),
    handleDeleteClick: jest.fn(),
    handleEditCallback: jest.fn(),
  };
  test('snapshot', () => expect(renderer.create(<PostListItem {...{ ...defaultProps }} />).toJSON()).toMatchSnapshot());
});
