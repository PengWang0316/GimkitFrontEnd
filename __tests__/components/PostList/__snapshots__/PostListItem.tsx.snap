// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostListItem snapshot 1`] = `
<List
  className="makeStyles-list-2"
>
  <ListItem
    alignItems="center"
    button={true}
    onClick={[Function]}
  >
    <ListItemText
      className="makeStyles-itemText-5"
      primary={
        <React.Fragment>
          <Typography
            className="makeStyles-inlineBold-4"
            color="textPrimary"
            component="span"
            variant="h6"
          >
            title - 1/1/2019
          </Typography>
          <IconButton
            className="makeStyles-editBtn-7"
            onClick={[Function]}
            size="small"
          >
            <EditIcon
              className="makeStyles-icon-8"
            />
          </IconButton>
          <IconButton
            className="makeStyles-deleteBtn-6"
            onClick={[Function]}
            size="small"
          >
            <DeleteIcon
              className="makeStyles-icon-8"
            />
          </IconButton>
        </React.Fragment>
      }
      secondary={
        <React.Fragment>
          <Typography
            className="makeStyles-inline-3"
            color="textPrimary"
            component="span"
            variant="body2"
          >
            <div>
              Parsed value
            </div>
          </Typography>
        </React.Fragment>
      }
    />
  </ListItem>
  <Divider
    component="li"
  />
</List>
`;
